### DISCUSSION TITLE
Cannot reconnect. Please reload the window. (nginx as a reverse proxy)

### DISCUSSION THREAD

**Discussion by nathanchu** [coder/code-server#2121] [Q&A]
I've just installed code-server on a DO droplet, but I am having some issues. I accessed it at http://[my-ip]:8080, and it seems to work fine. However, when I set up nginx and used this as a reverse proxy to port 8080, it only works for around a minute before I get a "Cannot reconnect. Please reload the window." error. I can reload it and it works fine, but it usually will stop working after a minute. I installed code-server via the script, with Node.js version 12.18.4 LTS. I don't know if this helps, but I have 1gb RAM and 1vCPU, so it may be an issue with the droplet being underpowered. I don't know if this helps either, but code-server got killed recently because of an OOME. If anyone could help, I'd really appreciate it.

**Comment by nathanchu** [comment]
I solved my own problem! I created a new droplet and it worked perfectly 🤷‍♂️. I guess I broke the config files.

**Comment by nathanchu** [comment]
Never mind, it seems that new VMs work better than older ones. I haven't edited the log files, but code-server pops up with "Cannot reconnect. Please reload the window." especially when opening large log files, and when opening the terminal. As a side note, this only happens when using nginx so it could either be the ssl cert or ngnix (I haven't tested nginx without ssl yet).

**Reply by code-asher** [reply]
Do you see any crash reports in the logs? https://github.com/cdr/code-server/blob/v3.5.0/doc/FAQ.md#how-do-i-debug-issues-with-code-server That should tell us if it's crashing due to an OOM.

**Reply by nathanchu** [reply]
It hasn't been giving any OOM errors anymore, and no meaningful errors in the debug logs, but every time I open the page I get the following errors in the console..

![image](https://user-images.githubusercontent.com/63111210/94497436-2ec85c80-01c5-11eb-8470-3530a07bb191.png)
(I've replaced my actual website with "website" to keep people from trying to brute force it)

If it would help I wouldn't mind upgrading. I want code-server to work but I don't want to buy more power than I need.

**EDIT:** Found this in remoteagent.log:
```
[remoteagent] [error] /root/.local/share/code-server/extensions {"errno":-2,"code":"ENOENT","syscall":"scandir","path":"/root/.local/share/code-server/extensions"}
```

**Reply by code-asher** [reply]
Hmm I don't think the extension host would crash if the extension directory didn't exist...or...maybe it would? If you create the directory does it help at all?

**Reply by nathanchu** [reply]
I just created it and still get the issue (without anything in the logs, but still the same errors in the chrome console). As a side note, I don't get the usual "Attempting to reconnect..." message. It just goes straight to "Cannot reconnect. Please reload the window."

**Reply by code-asher** [reply]
Are there any new errors in the log?

**Reply by code-asher** [reply]
In addition to `remote-agent.log` be sure to check `extension-host/exthost.log` although you're probably already aware. :smile:

**Reply by nathanchu** [reply]
> In addition to `remote-agent.log` be sure to check `extension-host/exthost.log` although you're probably already aware. 😄

Nothing in those files. If it would help I will upgrade my droplet to 2 vCPUs.

**Reply by code-asher** [reply]
It might be worth trying but I'm not sure more CPU will help. Do you see any `BUSY` lines in the logs? If so then it might be CPU.

**Reply by nathanchu** [reply]
> It might be worth trying but I'm not sure more CPU will help. Do you see any `BUSY` lines in the logs? If so then it might be CPU.

No, no busy lines.
I just looked at my DigitalOcean dashboard and the cpu is always under 50% so probably not an issue. Could this be an internet speed issue? It could also be with my config maybe:

```
server {

  server_name ide.mywebsite.com;

  location / {
    proxy_pass http://127.0.0.1:8080;
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection "Upgrade";
    proxy_set_header Host $host;
  }

    listen [::]:443 ssl ipv6only=on; # managed by Certbot
    listen 443 ssl; # managed by Certbot
    ssl_certificate /etc/letsencrypt/live/ide.mysebsite.com/fullchain.pem; # managed by Certbot
    ssl_certificate_key /etc/letsencrypt/live/ide.mywebsite.com/privkey.pem; # managed by Certbot
    include /etc/letsencrypt/options-ssl-nginx.conf; # managed by Certbot
    ssl_dhparam /etc/letsencrypt/ssl-dhparams.pem; # managed by Certbot

}
server {
    if ($host = ide.mywebsite.com) {
        return 301 https://$host$request_uri;
    } # managed by Certbot



  listen 80;
  listen [::]:80;

  server_name ide.mywebsite.com;
    return 404; # managed by Certbot


}
```
Obviously ide.mywebsite.com is a placeholder for my actual website.

**Reply by code-asher** [reply]
Hmmm, that configuration looks good to me. Although, you did say it only happens with nginx right? Would you be able to test http to see if it's something around TLS? And do you get the green lock in your browser?

If http doesn't work either then it might help to post the VS Code (the latest ones), code-server, and browser logs in their entirety in case something stands out to me.

**Reply by nathanchu** [reply]
> Hmmm, that configuration looks good to me. Although, you did say it only happens with nginx right? Would you be able to test http to see if it's something around TLS? And do you get the green lock in your browser?
> 
> If http doesn't work either then it might help to post the VS Code (the latest ones), code-server, and browser logs in their entirety in case something stands out to me.

I actually got a blank page with no https. I got the green lock before disabling https, but not after. I did change the file in ~/.config/code-server/config.yaml to 0.0.0.0 instead of 127.0.0.1, and interestingly enough, once I enabled https, it worked much better. This doesn't make sense though because nginx should just be taking incoming requests and sending them to 127.0.0.1, so it should have no difference. I still get the error when opening the terminal after switching folders, but not when opening large files. If you want I'll send you the logs, but if you wouldn't mind I'd like to send them to your email address listed on github because I'm sure some include my website name, password, and other potentially sensitive information.

**Reply by nathanchu** [reply]
> > Hmmm, that configuration looks good to me. Although, you did say it only happens with nginx right? Would you be able to test http to see if it's something around TLS? And do you get the green lock in your browser?
> > If http doesn't work either then it might help to post the VS Code (the latest ones), code-server, and browser logs in their entirety in case something stands out to me.
> 
> I actually got a blank page with no https. I got the green lock before disabling https, but not after. I did change the file in ~/.config/code-server/config.yaml to 0.0.0.0 instead of 127.0.0.1, and interestingly enough, once I enabled https, it worked much better. This doesn't make sense though because nginx should just be taking incoming requests and sending them to 127.0.0.1, so it should have no difference. I still get the error when opening the terminal after switching folders, but not when opening large files. If you want I'll send you the logs, but if you wouldn't mind I'd like to send them to your email address listed on github because I'm sure some include my website name, password, and other potentially sensitive information.

Never mind, it's still erroring when opening large files and just randomly. I'll send those logs to your email if it's okay with you.

**Reply by code-asher** [reply]
Yup that works for me. The VS Code logs should be clean but I think you're right that code-server's logs include the password if you're in debug mode which we should fix.

Although, thinking more about this problem it may be that the server logs won't help (the browser ones might). I think it might be that the websockets aren't connecting at all. If the page doesn't have a green lock that means the browser thinks the certificate is invalid and I think Chrome will not allow websocket connections with invalid certs.

That's why I was thinking that using http might work (since there would be no invalid certificates for Chrome to complain about) but if that's just showing a blank page while https doesn't then it seems like that might be an nginx configuration issue.

**Reply by nathanchu** [reply]
> Yup that works for me. The VS Code logs should be clean but I think you're right that code-server's logs include the password if you're in debug mode which we should fix.
> 
> Although, thinking more about this problem it may be that the server logs won't help (the browser ones might). I think it might be that the websockets aren't connecting at all. If the page doesn't have a green lock that means the browser thinks the certificate is invalid and I think Chrome will not allow websocket connections with invalid certs.
> 
> That's why I was thinking that using http might work (since there would be no invalid certificates for Chrome to complain about) but if that's just showing a blank page while https doesn't then it seems like that might be an nginx configuration issue.

It's actually a gray lock. I hope that's okay. The cert still appears trusted at ssllabs, and in my browser.

**Reply by code-asher** [reply]
I think as long as the browser logs don't have any errors about it you're good. For example here's what mine says when I load a certificate it doesn't like:

```
vscode.js:993 WebSocket connection to 'wss://8080.dev.localhost:44983/' failed: Error in connection establishment: net::ERR_CONNECTION_REFUSED
Uncaught (in promise) DOMException: Failed to register a ServiceWorker for scope ('https://8080.dev.localhost/') with script ('https://8080.dev.localhost/static/development/src/code-server/dist/serviceWorker.js'): An SSL certificate error occurred when fetching the script.
```

**Reply by nathanchu** [reply]
> I think as long as the browser logs don't have any errors about it you're good. For example here's what mine says when I load a certificate it doesn't like:
> 
> ```
> vscode.js:993 WebSocket connection to 'wss://8080.dev.localhost:44983/' failed: Error in connection establishment: net::ERR_CONNECTION_REFUSED
> Uncaught (in promise) DOMException: Failed to register a ServiceWorker for scope ('https://8080.dev.localhost/') with script ('https://8080.dev.localhost/static/development/src/code-server/dist/serviceWorker.js'): An SSL certificate error occurred when fetching the script.
> ```

I don't see those.  I just scanned through the logs again and it is, in fact, getting killed by OOM! Your system requirements say that 1gb ram should be fine and that's what I have though. I'm going to test this locally (with a vm, 2gb ram and 2vCPU to make sure OOM won't be a problem) and give myself a certificate signed by my own private CA (to better emulate let's encrypt), and trust that root certificate in chrome. I'll let you know if I still receive the error of cannot reconnect. I have seen the following error when running it locally though, so if it helps:
```
(node:31) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 exit listeners added to [ChildProcess]. Use emitter.setMaxListeners() to increase limit

(node:31) MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 error listeners added to [ChildProcess]. Use emitter.setMaxListeners() to increase limit
```

**Reply by code-asher** [reply]
Ahh I'm sorry about that. We pull the 1GB requirement from VS Code but it's possible code-server has a higher overhead or maybe there is a memory leak like noted in those logs. I'll investigate.
